### DISCUSSION TITLE
always getting a \r\n at the end of my string with StringWriter

### DISCUSSION THREAD

**Discussion by madquantum81** [JoshClose/CsvHelper#1781] [Q&A] [answered]
Hello,
When I create a csv string with a StringWriter, I always get a  \r\n at the end of that string.. Is there a way to remove it?

Heres my config

var csvConfig = new CsvConfiguration(CultureInfo.InvariantCulture)
            {
                HasHeaderRecord = false,
                TrimOptions = TrimOptions.Trim,
            };

Thanks!

**Comment by JoshClose** [comment, marked as answer]
How are you doing the writing?

**Reply by madquantum81** [reply]
var csvContent = new StringBuilder();
using (var writer = new StringWriter(csvContent))
            using (var csv = new CsvWriter(writer, csvConfig))
            {
                csv.Context.TypeConverterOptionsCache.GetOptions(typeof(DateTime)).Formats = new[] { "yyyy/MM/dd" };
                csv.WriteRecords(listData);
            }

**Reply by JoshClose** [reply]
`\r\n` is a record separator. It's a carriage return and a line feed. This is part of the RFC 4180 spec. If you hit `Enter` on your keyboard, a `\r\n` is what is generated from it (on Windows).

**Reply by madquantum81** [reply]
yeah but why its generating a \r\n after the last item? Is there a way to not have a carriage return a the end?

**Reply by JoshClose** [reply]
You'll have to write manually if you want to do that.

```cs
void Main()
{
    var records = new List<Foo>
    {
        new Foo { Id = 1, Name = "one" },
    };
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var writer = new StringWriter())
    using (var csv = new CsvWriter(writer, config))
    {
        csv.WriteHeader<Foo>();
        foreach (var record in records)
        {
            csv.NextRecord();
            csv.WriteRecord(record);
        }

        csv.Flush();
        writer.ToString().ToCharArray().Dump();
    }
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
}
```

![image](https://user-images.githubusercontent.com/114971/115615837-15981200-a2b5-11eb-8039-f8404b91a2c9.png)

**Reply by JoshClose** [reply]
`NextRecord` writes the line ending.

**Reply by madquantum81** [reply]
ok nice!

The /r/n now is at the beginning so I ended up with

for (int i = 0; i < listData.Count; i++)
{
          if (i > 0)
          {
              csv.NextRecord();
          }
          csv.WriteRecord(listData[i]);
}

working fine! thanks for your help!

**Reply by JoshClose** [reply]
Forgot you weren't writing the header. Great!
